fdbserver/TLogGroup.actor.cpp

```diff
@@ -0,0 +1,293 @@
+#include "fdbserver/WorkerInterface.actor.h"
+#include "flow/Arena.h"
+#include "flow/FastRef.h"
+#include "flow/flow.h"
+#include "flow/Error.h"
+#include "flow/UnitTest.h"
+#include "flow/serialize.h"
+#include <iostream>
+#include <iterator>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+TLogGroupCollection::TLogGroupCollection(const Reference<IReplicationPolicy>& policy, int numGroups, int groupSize)
+  : policy(policy), targetNumGroups(numGroups), GROUP_SIZE(groupSize) {}
+
+const std::vector<TLogGroupRef>& TLogGroupCollection::groups() const {
+	return recruitedGroups;
+}
+
+int TLogGroupCollection::groupSize() const {
+	return GROUP_SIZE;
+}
+
+void TLogGroupCollection::addWorkers(const std::vector<WorkerInterface>& logWorkers) {
+	for (const auto& worker : logWorkers) {
+		recruitMap.emplace(worker.id(), TLogWorkerData::fromInterface(worker));
+	}
+}
+
+void TLogGroupCollection::recruitEverything() {
+	std::unordered_set<UID> selectedServers;
+	std::vector<TLogWorkerData*> bestSet;
+	auto localityMap = buildLocalityMap(selectedServers);
+
+	printf("---> Going to recruit %d groups out of %d servers\n", targetNumGroups, localityMap.size());
+	while (recruitedGroups.size() < targetNumGroups) {
+		bestSet.clear();
+
+		// TODO: We are doing this randomly for now, but should make sure number of teams served by each
+		//   tLog server is approximately same.
+		if (localityMap.selectReplicas(policy, bestSet)) {
+			// ASSERT_WE_THINK(bestSet.size() == GROUP_SIZE);
+
+			printf("----> Recruiting Group: %d\n", bestSet.size());
+			Reference<TLogGroup> group(new TLogGroup());
+			for (auto& entry : bestSet) {
+				printf("Recruited Group: %s\n", entry->locality.processId().get().toString().c_str());
+				group->addServer(Reference<TLogWorkerData>(entry));
+				selectedServers.insert(entry->id);
+			}
+
+			newGroups.push_back(group);
+		} else {
+			// TODO: We may have scenarios (simulation), with recruits/zone's < RF. Handle that case.
+		}
+	}
+
+	// ASSERT_EQ(newGroups.size(), numTeams);
+}
+
+LocalityMap<TLogWorkerData> TLogGroupCollection::buildLocalityMap(const std::unordered_set<UID>& ignoreServers) {
+	LocalityMap<TLogWorkerData> localityMap;
+	for (const auto& [_, logInterf] : recruitMap) {
+		if (ignoreServers.find(logInterf->id) != ignoreServers.end()) {
+			// Server already selected.
+			continue;
+		}
+		localityMap.add(logInterf->locality, logInterf.getPtr());
+	}
+	return localityMap;
+}
+
+void TLogGroupCollection::storeState(CommitTransactionRequest* recoveryCommitReq) {
+	CommitTransactionRef& tr = recoveryCommitReq->transaction;
+	const auto& serversPrefix = LiteralStringRef("/servers");
+
+	tr.clear(recoveryCommitReq->arena, tLogGroupKeys);
+	for (const auto& group : recruitedGroups) {
+		const auto& groupPrefix = tLogGroupKeyFor(group->id());
+		std::cout << "---> Storing " << group->toString() << std::endl;
+		tr.set(recoveryCommitReq->arena, groupPrefix.withSuffix(serversPrefix), group->toValue());
```

Comment:
Maybe just define a key, e.g., `groupServerPrefix`, for `groupPrefix.withSuffix(serversPrefix)`?